### PR TITLE
unify `refresh.sh` and `extract.py` into `refresh.py`

### PR DESCRIPTION
**Pull request by lummax**
Let's keep it small and incremental, this is the first part and a preparation for ongoing effort to support:
- automatic replacing of `bazel-out/` and `external` references
- Windows

I am not happy to what I did to the docstrings in the first commit. But I am also weary as not put to much refactoring/reorganizing into this file at this point to make this PR reviewable.

Feel free to drop the last two commits if you don't like `black` or my comment reformating.

### CODE REVIEW

**Review by aminya — APPROVED** [review]
I tried this branch, it seems to fix #19 if I run `bazel run @hedron_compile_commands//:refresh_all --incompatible_load_python_rules_from_bzl=false`. But I also got some build errors. Not sure if that is related.

**Comment by cpsauer** [comment]
@aminya, I'm pretty sure that's just an artifact of this PR factoring the py_binary out of the BUILD file. Tried to give a more full answer in #19 about how the flags were interacting and how to fix. 

We should probably keep this PR independent. But way to be resourceful!

File: refresh.template.py
Comment (cpsauer) [inline refresh.template.py]: Maybe let's toss this down next to the argument list itself? I only didn't do it before because of Bash's rigidities.

File: refresh.template.py
Comment (cpsauer) [inline refresh.template.py:416]: Of the things here, I'm probably most concerned about what happens here when there are errors.

Ideally, we'd have the following behavior, I think:
- Still report errors from aquery to the user, ideally including warnings.
- Continue on when sub-targets fail.
- And shush the known/fine warnings parallel to how we previously did with sed in the shell script.
  -  Without this last one, we'll get lots of people unnecessarily worried and filing issues.

Perhaps the simplest, if not quite the best, would be printing stderr from a try block iff there was an error, and then continuing to the next command? And the best would be always printing a filtered stderr and catching errors with a try.

Either way I'd also love to preserve the note about the missing graph targets error message so it doesn't throw people for a loop.

File: refresh.template.py
Comment (cpsauer) [inline refresh.template.py:338]: I know you may disagree, but could we not hard-wrap this?
[Normally I'd just reformat rather than bother you.]

File: refresh.template.py
Comment (cpsauer) [inline refresh.template.py:372]: I know you may disagree, but could we not hard-wrap this?
[Normally I'd just reformat rather than bother you.]

File: refresh_compile_commands.bzl
Comment (cpsauer) [inline refresh_compile_commands.bzl:65]: Wahoo! Thank you!

**Review by cpsauer — COMMENTED** [review]
Lukas, thank you so much for a high quality contribution here. I really appreciate your rolling up your sleeves to make things better and more cross-platform--and how good you've been about communicating.

On my end, my sincere apologies for not responding with my usual speed. I know it's annoying to have these things stall--whatever the reason--and I'll try my best to do better in the future. I'm looking forward to working with you on getting this in.

PR feedback that fits better here than in a comment:
Mostly, I'm delighted. I really appreciate your work. There are a few changes I'd like to propose before we merge, though.

1. Normally I'd just toss up commits fixing the little things I happen to notice while reading (duplicate imports, etc.) rather than spending your time on them in comments, but I don't think I have edit access to the PR. 
    1. I think maybe that's a result of it coming from an org-owned repo, where GitHub doesn't even show the "allow maintainer access" checkbox, rather than a user-owned repo. If you'd like me to just clean up small things in future PR reviews by giving me access, I promise I'll be responsible about it. Obviously, if you disagree with any cleanups I propose by doing them, just say and we'll revert and discuss.
2. I appreciate your efforts to break things up. It was *so* nice to be able to be able to review the first commit as a rename-and-edit so I could easily see the diff. Thank you.
    1. That said, I do want to also say that I'm less concerned about commit-history cleanliness than saving you time on a high quality result. So feel free to just do whatever you think most efficient for us overall.
3. Thanks for your efforts to reformat and keep things clean. If it's okay, though, I'd really like to not switch to hard-wrapping-everywhere (comments especially), on the theory that we're all on variable width screens rather than fixed-width terminals/punchcards (the origin of the convention!). I get that it's not conventional, but if I could take you up on your offer to not make that change, I'd love it. That said, totally fine to multiline code when you think it's more readable.

Again, thank you so much! Enjoying getting to working with you.
Chris

File: refresh.template.py
Comment (lummax) [reply refresh.template.py]: We are actually running this on `3.7`.

> Any special reason to have it that I might not be aware of?

Just personal preference :grimacing:  .. With `typing` pulled in we don't need to touch the imports if we also need for example `typing.Callable` or `typing.DefaultDict`.

File: refresh_compile_commands.bzl
Comment (lummax) [reply refresh_compile_commands.bzl:56]: You are right. I was just concerned to keep it syntactically correct python. This does the job :+1:

File: refresh.template.py
Comment (cpsauer) [inline refresh.template.py:448]: Whoops! Just realized: Can we keep the old indent=2, check_circular=False, for readability and speed, respectively? 

[Bed time over here, but thanks so much for looking through!]

File: refresh_compile_commands.bzl
Comment (cpsauer) [reply refresh_compile_commands.bzl:56]: I went through the exact same thought process :) Thanks for being so thoughtful about it.

**Comment by cpsauer** [comment]
@lummax, at the risk of being annoying after having been slow myself:

Anything I can do to help things along here? I saw you'd replied some right away. And I'm assuming you just haven't gotten a chance to get back to it. But please, don't hesitate to say if there's something you'd like from me.

**Comment by lummax** [comment]
Sorry, this was just pushed down my TODO list as other stuff popped up. I hope I'll get to it this week.

I really appreciate your comments and will address them :)

**Comment by cpsauer** [comment]
Not to worry--and no pressure. Good luck with those other things! 

Thanks for your contributions, and sorry to be a bit of a picky monkey.

File: refresh.template.py
Comment (lummax) [reply refresh.template.py:16]: I am not sure I follow on "..and move the ... expansion parts of the old docstring.." so I'll leave that to you and later :)

File: refresh.template.py
Comment (lummax) [reply refresh.template.py]: Sorry, I am not sure what you mean. Split this up and distribute it into the `cmd` list?

File: refresh.template.py
Comment (lummax) [reply refresh.template.py:229]: Yes.

File: refresh.template.py
Comment (lummax) [reply refresh.template.py:302]: Done as a fixup.

File: refresh.template.py
Comment (lummax) [reply refresh.template.py:243]: Done.

File: refresh.template.py
Comment (lummax) [reply refresh.template.py:269]: Done

File: refresh.template.py
Comment (lummax) [reply refresh.template.py]: Done

File: refresh_compile_commands.bzl
Comment (lummax) [reply refresh_compile_commands.bzl:44]: 1) The shebang is just habit of always adding it to files. But it is not necessary here, you are right. I'll go with (i)! :+1: 
2) No, I am not super convinced. In fact I'd rather be able to to `bazel run :extract //my:target`. But that is another PR I guess :)

Generating a JSON is a good idea. I was just lazy in the scope of this PR and opted to keep it similar to the previous approach.

File: refresh_compile_commands.bzl
Comment (lummax) [reply refresh_compile_commands.bzl:50]: Done.

File: refresh.template.py
Comment (lummax) [reply refresh.template.py:346]: :+1: on the rename.

**Comment by lummax** [comment]
Thank you for doing the remaining work. :tada:

**Comment by cpsauer** [comment]
No, thank **you**, for moving us in a great new direction, away from bash and towards platform independence. 🎉

Looking forward to taking the next steps with you!

**Pull request merged**
